### ISSUE TITLE
5.8.2 cores or loops using hashref as subscript

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7151]
Migrated from [rt.perl.org#27392](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=27392) (status was 'resolved')

Searchable as RT27392$

**Comment by p5pRT** [comment]
### From keni@aol.net

#### Created by keni@aol.net

This is a bug report for perl from keni@&#8203;aol\.net\,
generated with the help of perlbug 1\.34 running under perl v5\.8\.2\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
The \(buggy\, I\-didn't\-mean\-to\-type\-that\) sample code below will core under
HPUX 11&#8203;:
&nbsp;    X&#8203;: 1
&nbsp;    X&#8203;: HASH\(0x4001ecc8\)
&nbsp;    Segmentation fault \(core dumped\)

gdb says&#8203;:
\#0  0x00058948 in Perl\_av\_extend \(av=0x4001208c\, key=1073840232\) at av\.c&#8203;:158
158                         ary\[\-\-tmp\] = &PL\_sv\_undef;
\(gdb\) where
\#0  0x00058948 in Perl\_av\_extend \(av=0x4001208c\, key=1073840232\) at av\.c&#8203;:158
\#1  0x00058f1c in Perl\_av\_store \(av=0x4001208c\, key=1073840232\, val=0x40018074\)
&nbsp;    at av\.c&#8203;:315
\#2  0x00058c68 in Perl\_av\_fetch \(av=0x4001208c\, key=1073840232\, lval=1\)
&nbsp;    at av\.c&#8203;:246
\#3  0x000b5120 in Perl\_pp\_aelem \(\) at pp\_hot\.c&#8203;:2974
\#4  0x000a7554 in Perl\_runops\_debug \(\) at dump\.c&#8203;:1438
\#5  0x0002ed5c in S\_run\_body \(oldscope=1\) at perl\.c&#8203;:1860
\#6  0x0002e8b8 in perl\_run \(my\_perl=0x11b000\) at perl\.c&#8203;:1779
\#7  0x0002a4cc in main \(argc=2\, argv=0x7eff0694\, env=0x11aea4\) at perlmain\.c&#8203;:86
\#8  0xc013e8fc in ?? \(\)
\#9  0x0005887c in Perl\_av\_extend \(av=0x7eff06a0\, key=2130642832\) at av\.c&#8203;:118
Cannot access memory at address 0xffffff6e

5\.6\.1 on the same HPUX box does the same\.

Under Mac OS X 10\.2\.8\, 5\.8\.2 it doesn't core\.  Instead it prints&#8203;:
&nbsp;    X&#8203;: 1
&nbsp;    X&#8203;: HASH\(0x28b224\)
&nbsp;    X&#8203;:
\(the last line repeated 7 million times before I killed it\)\.

5\.6\.0 \(the official apple binary\) also goes into a loop\.

Sample code&#8203;:

$G\->\{A\}\{B\}\[0\] = 1;
\# what I meant&#8203;:
\#       foreach \(0\.\.$\#\{$G\->\{A\}\{B\}\} \)\{
\# what I wrote\, causes core or loop&#8203;:
&nbsp;    foreach \(@&#8203;\{$G\->\{A\}\{B\}\} \)\{
&nbsp;        print "X&#8203;: $\_\\n";
&nbsp;        $G\->\{A\}\{B\}\[$\_\]\{C\};
&nbsp;    \}

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.2:

Configured by tsadm at Fri Feb 13 13:59:59 EST 2004.

Summary of my perl5 (revision 5.0 version 8 subversion 2) configuration:
  Platform:
    osname=hpux, osvers=11.00, archname=PA-RISC2.0
    uname='hp-ux tsdev-m1 b.11.00 a 9000800 633329363 two-user license '
    config_args='-Dprefix=/stats/process2/stats_share -Dcc=/usr/cygnus/bin/gcc -DPERL_POLLUTE_MALLOC'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='/usr/cygnus/bin/gcc', ccflags ='-D_HPUX_SOURCE -DUINT32_MAX_BROKEN -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g -O',
    cppflags='-D__STDC_EXT__ -D_HPUX_SOURCE -D_HPUX_SOURCE -DUINT32_MAX_BROKEN -DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.9-gnupro-98r2', gccosandvers='hpux11.00'
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='/usr/bin/ld', ldflags =' -Wl,+vnocompatwarnings -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/ccs/lib
    libs=-lnsl -lnm -lndbm -lgdbm -ldb -ldld -lm -lcrypt -lsec -lc
    perllibs=-lnsl -lnm -ldld -lm -lcrypt -lsec -lc
    libc=/lib/libc.sl, so=sl, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_hpux.xs, dlext=sl, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-B,deferred '
    cccdlflags='-fPIC', lddlflags='-b -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.2:
    /stats/process2/stats_share/lib/perl5/5.8.2/PA-RISC2.0
    /stats/process2/stats_share/lib/perl5/5.8.2
    /stats/process2/stats_share/lib/perl5/site_perl/5.8.2/PA-RISC2.0
    /stats/process2/stats_share/lib/perl5/site_perl/5.8.2
    /stats/process2/stats_share/lib/perl5/site_perl
    .


Environment for perl v5.8.2:
    HOME=/home/keni
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=.share/bin:/opt/perl5/bin:/home/keni/bin:/cm/bin/cvs/bin:/usr/cygnus/bin:/usr/bin:/opt/ansic/bin:/usr/ccs/bin:/usr/contrib/bin:/opt/nettladm/bin:/opt/fc/bin:/opt/fcms/bin:/opt/upgrade/bin:/opt/pd/bin:/usr/bin/X11:/usr/contrib/bin/X11:/opt/hparray/bin:/opt/resmon/bin:/opt/pred/bin:/opt/perf/bin:/opt/OV/bin/OpC:/opt/cxperf/bin:/opt/langtools/bin:/opt/imake/bin:/opt/graphics/common/bin:/usr/sbin/diag/contrib:/opt/aCC/bin:/usr/local/bin
    PERL_BADLANG (unset)
    SHELL=/usr/bin/csh


```
</details>

**Comment by p5pRT** [comment]
### From brentdax@cpan.org

>     foreach \(@&#8203;\{$G\->\{A\}\{B\}\} \)\{
>         print "X&#8203;: $\_\\n";
>         $G\->\{A\}\{B\}\[$\_\]\{C\};
>     \}

The hashref is being evaluated as a number\, which returns the numeric 
address of the referee\.

On HP\-UX\, that means you're trying to access an array element with index 
1\,073\,867\,976\.  Presumably\, HP\-UX is refusing to allocate 1\,073\,867\,977 
scalars\, making ary\[1\,073\,867\,976\] \(or thereabouts\) a bad pointer\.  \(I 
don't know whether 'ary' is null or not\, and I'm not really in a 
position to check\-\-I don't have a Perl with debug symbols handy\.\)

On OS X\, you're trying to access element 2\,667\,044\.  Apparently this is 
okay on your box \(although I'm guessing Perl's eating a few extra 
megabytes\)\.  However\, note the caveat in L\<perlsyn/Foreach Loops>&#8203;:

&nbsp;    If any part of LIST is an array\, foreach will get very confused
&nbsp;    if you add or remove elements within the loop body\, for example
&nbsp;    with splice\. So don't do that\.

That's good advice on both of these problems\, really\-\-don't do that\.

\-\- 
Brent "Dax" Royal\-Gordon \<brent@&#8203;brentdax\.com>
Perl and Parrot hacker

Oceania has always been at war with Eastasia\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @timbunce

On Thu\, Mar 04\, 2004 at 06&#8203;:15&#8203;:02PM \-0000\, Kenneth Lorber wrote&#8203;:
> HPUX 11&#8203;:
>     X&#8203;: 1
>     X&#8203;: HASH\(0x4001ecc8\)
>     Segmentation fault \(core dumped\)
> 
> gdb says&#8203;:
> \#0  0x00058948 in Perl\_av\_extend \(av=0x4001208c\, key=1073840232\) at av\.c&#8203;:158
> 158                         ary\[\-\-tmp\] = &PL\_sv\_undef;

The bug is that the failed memory allocation for 1073840232\*sizeof\(SV\*\)
wasn't detected\.

> 5\.6\.1 on the same HPUX box does the same\.
> 
> Under Mac OS X 10\.2\.8\, 5\.8\.2 it doesn't core\.  Instead it prints&#8203;:
>     X&#8203;: 1
>     X&#8203;: HASH\(0x28b224\)
>     X&#8203;:
> \(the last line repeated 7 million times before I killed it\)\.
> 
> 5\.6\.0 \(the official apple binary\) also goes into a loop\.

0x28b224 is a lot smaller than 0x4001ecc8\. The memory allocation worked
and it's printing millions of undefs\.

Tim\.

> Sample code&#8203;:
> 
> $G\->\{A\}\{B\}\[0\] = 1;
> \# what I meant&#8203;:
> \#       foreach \(0\.\.$\#\{$G\->\{A\}\{B\}\} \)\{
> \# what I wrote\, causes core or loop&#8203;:
>     foreach \(@&#8203;\{$G\->\{A\}\{B\}\} \)\{
>         print "X&#8203;: $\_\\n";
>         $G\->\{A\}\{B\}\[$\_\]\{C\};
>     \}
> 
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=low
> \-\-\-
> Site configuration information for perl v5\.8\.2&#8203;:
> 
> Configured by tsadm at Fri Feb 13 13&#8203;:59&#8203;:59 EST 2004\.
> 
> Summary of my perl5 \(revision 5\.0 version 8 subversion 2\) configuration&#8203;:
>   Platform&#8203;:
>     osname=hpux\, osvers=11\.00\, archname=PA\-RISC2\.0
>     uname='hp\-ux tsdev\-m1 b\.11\.00 a 9000800 633329363 two\-user license '
>     config\_args='\-Dprefix=/stats/process2/stats\_share \-Dcc=/usr/cygnus/bin/gcc \-DPERL\_POLLUTE\_MALLOC'
>     hint=recommended\, useposix=true\, d\_sigaction=define
>     usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
>     useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
>     use64bitint=undef use64bitall=undef uselongdouble=undef
>     usemymalloc=y\, bincompat5005=undef
>   Compiler&#8203;:
>     cc='/usr/cygnus/bin/gcc'\, ccflags ='\-D\_HPUX\_SOURCE \-DUINT32\_MAX\_BROKEN \-DDEBUGGING \-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
>     optimize='\-g \-O'\,
>     cppflags='\-D\_\_STDC\_EXT\_\_ \-D\_HPUX\_SOURCE \-D\_HPUX\_SOURCE \-DUINT32\_MAX\_BROKEN \-DDEBUGGING \-fno\-strict\-aliasing \-I/usr/local/include'
>     ccversion=''\, gccversion='2\.9\-gnupro\-98r2'\, gccosandvers='hpux11\.00'
>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=4321
>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=8
>     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
>     alignbytes=8\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='/usr/bin/ld'\, ldflags =' \-Wl\,\+vnocompatwarnings \-L/usr/local/lib'
>     libpth=/usr/local/lib /lib /usr/lib /usr/ccs/lib
>     libs=\-lnsl \-lnm \-lndbm \-lgdbm \-ldb \-ldld \-lm \-lcrypt \-lsec \-lc
>     perllibs=\-lnsl \-lnm \-ldld \-lm \-lcrypt \-lsec \-lc
>     libc=/lib/libc\.sl\, so=sl\, useshrplib=false\, libperl=libperl\.a
>     gnulibc\_version=''
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_hpux\.xs\, dlext=sl\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E \-Wl\,\-B\,deferred '
>     cccdlflags='\-fPIC'\, lddlflags='\-b \-L/usr/local/lib'
> 
> Locally applied patches&#8203;:
>     
> 
> \-\-\-
> @&#8203;INC for perl v5\.8\.2&#8203;:
>     /stats/process2/stats\_share/lib/perl5/5\.8\.2/PA\-RISC2\.0
>     /stats/process2/stats\_share/lib/perl5/5\.8\.2
>     /stats/process2/stats\_share/lib/perl5/site\_perl/5\.8\.2/PA\-RISC2\.0
>     /stats/process2/stats\_share/lib/perl5/site\_perl/5\.8\.2
>     /stats/process2/stats\_share/lib/perl5/site\_perl
>     \.
> 
> \-\-\-
> Environment for perl v5\.8\.2&#8203;:
>     HOME=/home/keni
>     LANG \(unset\)
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>     PATH=\.share/bin&#8203;:/opt/perl5/bin&#8203;:/home/keni/bin&#8203;:/cm/bin/cvs/bin&#8203;:/usr/cygnus/bin&#8203;:/usr/bin&#8203;:/opt/ansic/bin&#8203;:/usr/ccs/bin&#8203;:/usr/contrib/bin&#8203;:/opt/nettladm/bin&#8203;:/opt/fc/bin&#8203;:/opt/fcms/bin&#8203;:/opt/upgrade/bin&#8203;:/opt/pd/bin&#8203;:/usr/bin/X11&#8203;:/usr/contrib/bin/X11&#8203;:/opt/hparray/bin&#8203;:/opt/resmon/bin&#8203;:/opt/pred/bin&#8203;:/opt/perf/bin&#8203;:/opt/OV/bin/OpC&#8203;:/opt/cxperf/bin&#8203;:/opt/langtools/bin&#8203;:/opt/imake/bin&#8203;:/opt/graphics/common/bin&#8203;:/usr/sbin/diag/contrib&#8203;:/opt/aCC/bin&#8203;:/usr/local/bin
>     PERL\_BADLANG \(unset\)
>     SHELL=/usr/bin/csh

**Comment by p5pRT** [comment]
### From keni@aol.net

&nbsp;    |From rt\-brent=brentdax\.com@&#8203;perl\.org Thu Mar  4 23&#8203;:33&#8203;:53 EST 2004
&nbsp;    |Date&#8203;: 5 Mar 2004 04&#8203;:33&#8203;:51 \-0000
&nbsp;    |Subject&#8203;: Re&#8203;: \[perl \#27392\] 5\.8\.2 cores or loops using hashref as subscript
&nbsp;    |From&#8203;: "Brent Dax via RT" \<perlbug\-followup@&#8203;perl\.org>
&nbsp;    |Reply\-To&#8203;: perlbug\-followup@&#8203;perl\.org
&nbsp;    |In\-Reply\-To&#8203;: \<rt\-27392@&#8203;perl>
&nbsp;    |Message\-ID&#8203;: \<rt\-3\.0\.8\-27392\-80833\.1\.2941310206859@&#8203;perl\.org>
&nbsp;    |X\-RT\-Loop\-Prevention&#8203;: perl
&nbsp;    |RT\-Ticket&#8203;: perl \#27392
&nbsp;    |RT\-Originator&#8203;: brent@&#8203;brentdax\.com
&nbsp;    |To&#8203;: keni@&#8203;aol\.net

&nbsp;    >     foreach \(@&#8203;\{$G\->\{A\}\{B\}\} \)\{
&nbsp;    >         print "X&#8203;: $\_\\n";
&nbsp;    >         $G\->\{A\}\{B\}\[$\_\]\{C\};
&nbsp;    >     \}

&nbsp;    The hashref is being evaluated as a number\, which returns the numeric 
&nbsp;    address of the referee\.
Agreed\.

&nbsp;    On HP\-UX\, that means you're trying to access an array element with index 
&nbsp;    1\,073\,867\,976\.  Presumably\, HP\-UX is refusing to allocate 1\,073\,867\,977 
&nbsp;    scalars\, making ary\[1\,073\,867\,976\] \(or thereabouts\) a bad pointer\.  \(I 
&nbsp;    don't know whether 'ary' is null or not\, and I'm not really in a 
&nbsp;    position to check\-\-I don't have a Perl with debug symbols handy\.\)
According to the stack trace I sent\, it's not an allocation issue\.  The core
is occuring in the loop that's trying to assign undef to the newly created
elements; this sounds to me like there's a missing bounds check\.

Tracing through with gdb\, Perl\_av\_extend sees a key of 1073840268 at av\.c&#8203;:81\.
At line 110\, we get newmax=4 \(MYMALLOC is defined in our perl build\)\.
At 115\, newmax=1073840268\.
But then 118 calls Renew\, which calculates the size of the memory block it
needs&#8203;: \(handy\.h\) \(MEM\_SIZE\)\(\(n\)\*sizeof\(t\)\) \- and we get an integer overflow
\(1073840268\+1\) \* 4 => 393780 \(after 32bit truncation\)\.

I think that's the bug \- perl thinks it allocated more memory than it did
and when it tries to undef the new elements\, boom\.  We need an overflow test
in \(or before\) the call to Renew\.  IMHO\.

&nbsp;    On OS X\, you're trying to access element 2\,667\,044\.  Apparently this is 
&nbsp;    okay on your box \(although I'm guessing Perl's eating a few extra 
&nbsp;    megabytes\)\.
Yes\, but I'd expect it to terminate after 2\,667\,044\, but in this case
the caveat you quoted below does apply\.

&nbsp;    However\, note the caveat in L\<perlsyn/Foreach Loops>&#8203;:

&nbsp;        If any part of LIST is an array\, foreach will get very confused
&nbsp;        if you add or remove elements within the loop body\, for example
&nbsp;        with splice\. So don't do that\.

&nbsp;    That's good advice on both of these problems\, really\-\-don't do that\.
I agree \- my code was in error\.  But I don't like being able to core an
interpreter \(and not get any indication of why\)\.

&nbsp;    Brent "Dax" Royal\-Gordon \<brent@&#8203;brentdax\.com>
&nbsp;    Perl and Parrot hacker
Thanks\,
keni

**Comment by p5pRT** [comment]
### From @jkeenan

I reviewed this older ticket tonight\.  Trimming the last post&#8203;:

\#\#\#\#\#
On Fri Mar 05 04&#8203;:19&#8203;:24 2004\, keni@&#8203;aol\.net wrote&#8203;:
\.\.\.
> I agree \- my code was in error\.  But I don't like being able to core
> an interpreter \(and not get any indication of why\)\.
> 

So it appears the OP concedes he was attempting something he should not
have\.  When I run the attached program \-\- which throws 'use strict' and
'use warnings' around the OP's code \-\- I get&#8203;:

\#\#\#\#\#
$ perl 27393\.pl 
Useless use of hash element in void context at 27393\.pl line 10\.
X&#8203;: 1
X&#8203;: HASH\(0x1804180\)
Use of reference "HASH\(0x1804180\)" as array index at 27393\.pl line 10\.
Use of uninitialized value $\_ in concatenation \(\.\) or string at 27393\.pl
line 9\.
X&#8203;: 
Use of uninitialized value $\_ in array element at 27393\.pl line 10\.
Can't use string \("1"\) as a HASH ref while "strict refs" in use at
27393\.pl line 10\.
\#\#\#\#\#

So I am amply warned of my bad behavior\.  I recommend we close this ticket\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

[27393.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/1070504/531996/27393.pl)

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Jan 27\, 2012 at 07&#8203;:50&#8203;:25PM \-0800\, James E Keenan via RT wrote&#8203;:
> I reviewed this older ticket tonight\.  Trimming the last post&#8203;:
> 
> \#\#\#\#\#
> On Fri Mar 05 04&#8203;:19&#8203;:24 2004\, keni@&#8203;aol\.net wrote&#8203;:
> \.\.\.
> > I agree \- my code was in error\.  But I don't like being able to core
> > an interpreter \(and not get any indication of why\)\.
> > 
> 
> So it appears the OP concedes he was attempting something he should not
> have\.  When I run the attached program \-\- which throws 'use strict' and
> 'use warnings' around the OP's code \-\- I get&#8203;:
> 
> \#\#\#\#\#
> $ perl 27393\.pl 
> Useless use of hash element in void context at 27393\.pl line 10\.
> X&#8203;: 1
> X&#8203;: HASH\(0x1804180\)
> Use of reference "HASH\(0x1804180\)" as array index at 27393\.pl line 10\.
> Use of uninitialized value $\_ in concatenation \(\.\) or string at 27393\.pl
> line 9\.
> X&#8203;: 
> Use of uninitialized value $\_ in array element at 27393\.pl line 10\.
> Can't use string \("1"\) as a HASH ref while "strict refs" in use at
> 27393\.pl line 10\.
> \#\#\#\#\#
> 
> So I am amply warned of my bad behavior\.  I recommend we close this ticket\.

I disagree\. The OP pointed out that the code segfaulted\. Bad perl code
shouldn't segfault\. There appear to be a couple of possible issues;

The first is modifying \(extending\) the array be looped over; this has
traditionally been regarded as a Bad Thing to do\, although I think we
handle this now\.

The second issue is that doing $array\[$ref\] numfies the address of $ref
and so indexes a huge array element\. I think the OP's code was segfaulting
when doing address calculations that overflowed a 32\-bit address\.
IIRC\, we've since hardened the array indexing code to be more robust against
this sort of thing\.

So\, I think we've \*probably\* fixed this\. but it would be nice if someone
could bisect this\.

It might be as simple as\, doing on a 32\-bit system\)\, $a\[1073840232\]
\(based on the OP's stack trace\)\.

\-\- 
You never really learn to swear until you learn to drive\.

**Comment by p5pRT** [comment]
### From @khwilliamson

On 01/29/2012 01&#8203;:56 PM\, Dave Mitchell wrote&#8203;:
> On Fri\, Jan 27\, 2012 at 07&#8203;:50&#8203;:25PM \-0800\, James E Keenan via RT wrote&#8203;:
>> I reviewed this older ticket tonight\.  Trimming the last post&#8203;:
>>
>> \#\#\#\#\#
>> On Fri Mar 05 04&#8203;:19&#8203;:24 2004\, keni@&#8203;aol\.net wrote&#8203;:
>> \.\.\.
>>> I agree \- my code was in error\.  But I don't like being able to core
>>> an interpreter \(and not get any indication of why\)\.
>>>
>>
>> So it appears the OP concedes he was attempting something he should not
>> have\.  When I run the attached program \-\- which throws 'use strict' and
>> 'use warnings' around the OP's code \-\- I get&#8203;:
>>
>> \#\#\#\#\#
>> $ perl 27393\.pl
>> Useless use of hash element in void context at 27393\.pl line 10\.
>> X&#8203;: 1
>> X&#8203;: HASH\(0x1804180\)
>> Use of reference "HASH\(0x1804180\)" as array index at 27393\.pl line 10\.
>> Use of uninitialized value $\_ in concatenation \(\.\) or string at 27393\.pl
>> line 9\.
>> X&#8203;:
>> Use of uninitialized value $\_ in array element at 27393\.pl line 10\.
>> Can't use string \("1"\) as a HASH ref while "strict refs" in use at
>> 27393\.pl line 10\.
>> \#\#\#\#\#
>>
>> So I am amply warned of my bad behavior\.  I recommend we close this ticket\.
>
> I disagree\. The OP pointed out that the code segfaulted\. Bad perl code
> shouldn't segfault\. There appear to be a couple of possible issues;
>
> The first is modifying \(extending\) the array be looped over; this has
> traditionally been regarded as a Bad Thing to do\, although I think we
> handle this now\.
>
> The second issue is that doing $array\[$ref\] numfies the address of $ref
> and so indexes a huge array element\. I think the OP's code was segfaulting
> when doing address calculations that overflowed a 32\-bit address\.
> IIRC\, we've since hardened the array indexing code to be more robust against
> this sort of thing\.
>
> So\, I think we've \*probably\* fixed this\. but it would be nice if someone
> could bisect this\.
>
> It might be as simple as\, doing on a 32\-bit system\)\, $a\[1073840232\]
> \(based on the OP's stack trace\)\.
>

I run on a 32 bit system\, and I ran this
perl5\.8\.9 \-e 'print $a\[1073840232\]'
Name "main&#8203;::a" used only once&#8203;: possible typo at \-e line 1\.
Use of uninitialized value in print at \-e line 1\.

which is the earliest perl version I have\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Jan 29 13&#8203;:10&#8203;:42 2012\, public@&#8203;khwilliamson\.com wrote&#8203;:
> On 01/29/2012 01&#8203;:56 PM\, Dave Mitchell wrote&#8203;:
> > So\, I think we've \*probably\* fixed this\. but it would be nice if
> someone
> > could bisect this\.
> >
> > It might be as simple as\, doing on a 32\-bit system\)\, $a\[1073840232\]
> > \(based on the OP's stack trace\)\.
> >
> 
> I run on a 32 bit system\, and I ran this
> perl5\.8\.9 \-e 'print $a\[1073840232\]'
> Name "main&#8203;::a" used only once&#8203;: possible typo at \-e line 1\.
> Use of uninitialized value in print at \-e line 1\.
> 
> which is the earliest perl version I have\.

\(I think you actually used \-we\.\)

I get the same results with 5\.6\.2 and 5\.8\.1\.  So that example is not
sufficient to reproduce the bug\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @nwc10

On Sun\, Jan 29\, 2012 at 08&#8203;:56&#8203;:49PM \+0000\, Dave Mitchell wrote&#8203;:
> I disagree\. The OP pointed out that the code segfaulted\. Bad perl code
> shouldn't segfault\. There appear to be a couple of possible issues;
> 
> The first is modifying \(extending\) the array be looped over; this has
> traditionally been regarded as a Bad Thing to do\, although I think we
> handle this now\.
> 
> The second issue is that doing $array\[$ref\] numfies the address of $ref
> and so indexes a huge array element\. I think the OP's code was segfaulting
> when doing address calculations that overflowed a 32\-bit address\.
> IIRC\, we've since hardened the array indexing code to be more robust against
> this sort of thing\.
> 
> So\, I think we've \*probably\* fixed this\. but it would be nice if someone
> could bisect this\.
> 
> It might be as simple as\, doing on a 32\-bit system\)\, $a\[1073840232\]
> \(based on the OP's stack trace\)\.

I'm finding it really hard to replicate the OP's problem\. The memory map
on Linux \(at least x86\)\, FreeBSD \(x86\)\, OS X \(x86\) and AIX \(Power?\) doesn't
fail in the right way\.

I have this suspicion that ia64 Linux would sort of be right\, except that
it's \(obviously\) not 32 bit\. :\-\)

$ perl \-MConfig \-le 'print "$Config\{archname\}&#8203;: " \. \[\]'
ia64\-linux\-gnu\-thread\-multi&#8203;: ARRAY\(0x6000000000024120\)

Note how the heap isn't at the bottom of address space\.


The OP was on HP/UX\. I don't have access to HP/UX

Nicholas Clark

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Feb 01\, 2012 at 09&#8203;:37&#8203;:49PM \+0000\, Nicholas Clark wrote&#8203;:
> I'm finding it really hard to replicate the OP's problem\. The memory map
> on Linux \(at least x86\)\, FreeBSD \(x86\)\, OS X \(x86\) and AIX \(Power?\) doesn't
> fail in the right way\.
> 
> I have this suspicion that ia64 Linux would sort of be right\, except that
> it's \(obviously\) not 32 bit\. :\-\)
> 
> $ perl \-MConfig \-le 'print "$Config\{archname\}&#8203;: " \. \[\]'
> ia64\-linux\-gnu\-thread\-multi&#8203;: ARRAY\(0x6000000000024120\)
> 
> Note how the heap isn't at the bottom of address space\.

I can reproduce it on 32\-bit x86\, 5\.8\.2&#8203;:

&nbsp;    $ \./perl \-e '$a\[1073840232\]=1'
&nbsp;    Segmentation fault \(core dumped\)

It dies in av\_extend\(\)\, due to trying to malloc slightly more than
&nbsp;    2^30 \* sizeof\(SV\*\)
and the calculated size wraps round to something small\, which then causes
the SEGV while trying to fill the array with &PL\_sv\_undef\.

This has been fixed by the MEM\_WRAP\_CHECK mechanism added in 5\.8\.4\, so I'm
closing the ticket\.

&nbsp;    $ \./perl \-e '$a\[1073840232\]=1'
&nbsp;    panic&#8203;: array extend at \-e line 1\.

\-\- 
"Do not dabble in paradox\, Edward\, it puts you in danger of fortuitous wit\."
&nbsp;    \-\- Lady Croom\, "Arcadia"

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
